sdk/tests/sdk/whirlpools/whirlpool-impl#collectFeesAndRewardsForPositions.test.ts

```diff
@@ -0,0 +1,402 @@
+          },
+          {
+            emissionsPerSecondX64: MathUtil.toX64(new Decimal(10)),
+            vaultAmount: new u64(vaultStartBalance),
+          },
+        ],
+      });
+
+      fixtures.push(fixture);
+      positions.push(...fixture.getInfos().positions);
+    }
+
+    await sleep(2); // accrueRewards
+    for (const fixture of fixtures) {
+      await accrueFees(fixture);
+      await (ataExists ? createATAs : burnAndCloseATAs)(fixture);
+      await stopRewardsEmission(fixture);
+    }
+
+    // check all positions have fees and rewards
+    for (const positionInfo of positions) {
+      const position = await testCtx.whirlpoolClient.getPosition(positionInfo.publicKey);
+
+      const poolData = await testCtx.whirlpoolCtx.fetcher.getPool(position.getData().whirlpool, true);
+      const positionData = await position.refreshData();
+      const tickLowerData = position.getLowerTickData();
+      const tickUpperData = position.getLowerTickData();
+
+      const feeQuote = collectFeesQuote({
+        whirlpool: poolData!,
+        position: positionData,
+        tickLower: tickLowerData,
+        tickUpper: tickUpperData,
+      });
+
+      const rewardQuote = collectRewardsQuote({
+        whirlpool: poolData!,
+        position: positionData,
+        tickLower: tickLowerData,
+        tickUpper: tickUpperData,
+      });
+
+      assert.ok(feeQuote.feeOwedA.gt(ZERO));
+      assert.ok(feeQuote.feeOwedB.gt(ZERO));
+      assert.ok(rewardQuote[0]?.gt(ZERO));
+      assert.ok(rewardQuote[1]?.gt(ZERO));
+      assert.ok(rewardQuote[2]?.gt(ZERO));
+    }
+
+    const txs = await testCtx.whirlpoolClient.collectFeesAndRewardsForPositions(
+      positions.map((p) => p.publicKey),
+      true,
+    );
+    assert.ok(txs.length >= 2);
+
+    const requests: SendTxRequest[] = [];
+    for (const tx of txs) {
+      requests.push(await tx.build());
+    }
+
+    const parallel = true;
+    const processor = new TransactionProcessor(testCtx.whirlpoolCtx.connection, testCtx.whirlpoolCtx.wallet);
+    const { execute } = await processor.signAndConstructTransactions(requests, parallel);
+
+    const txResults = await execute();
+    for (const result of txResults) {
+      assert.ok(result.status === "fulfilled");
+    }
+
+    // check all positions have no fees and rewards
+    for (const positionInfo of positions) {
+      const position = await testCtx.whirlpoolClient.getPosition(positionInfo.publicKey);
+
+      const poolData = await testCtx.whirlpoolCtx.fetcher.getPool(position.getData().whirlpool, true);
+      const positionData = await position.refreshData();
+      const tickLowerData = position.getLowerTickData();
+      const tickUpperData = position.getLowerTickData();
+
+      const feeQuote = collectFeesQuote({
+        whirlpool: poolData!,
+        position: positionData,
+        tickLower: tickLowerData,
+        tickUpper: tickUpperData,
+      });
+
+      const rewardQuote = collectRewardsQuote({
```

Comment:
pass in `timeStampInSeconds: pool.getData().rewardLastUpdatedTimestamp` here.
CollectRewardQuote needs the latest timestamp on-chain to derive the correct derived reward values on-chain. (should be fine since you are not checking exact numbers, but still good to have it just in case)

Reply:
makes a lot of sense!
I have added timeStampInSeconds param. 🙂

Reply:
@atamari-orca 
Thanks for your review!

nope.
Since rewardLastUpdatedTimestamp is not updated until a transaction occurs, specifying rewardLastUpdatedTimestamp would underestimate the value in a pool where rewards are being paid out but no swap has occurred at all.

In testing, it makes sense to check the exact value based on rewards paid out up to a certain point in time.
On the other hand, a script examining pending reward should also take into account the passage of time up to now.
Since omitting this argument uses the current time, I believe it is correct to omit it and use the default value.